sympy/functions/combinatorial/factorials.py

```diff
@@ -866,192 +957,542 @@ def fdiff(self, argindex=1):
+
+        # n < 0 <= k
+        if n.is_negative and k.is_nonnegative:
+            if kint:
+                n = -n + k - 1
                 res = cls._eval(n, k)
-                return res.expand(basic=True) if res else res
-        elif n_nonneg is False and n_isint:
-            # a special case when binomial evaluates to complex infinity
-            return S.ComplexInfinity
-        elif k.is_number:
-            from sympy import gamma
-            return gamma(n + 1)/(gamma(k + 1)*gamma(n - k + 1))
+                return res if res is None else Pow(S.NegativeOne, k)*res
+            # if k is not an integer then we have n = neq, k=pos
+            # gamma(neg + 1)/(gamma(pos + 1)*gamma(neg - pos + 1))
+            # and the only gamma that will evaluate is that which is an
+            # odd multiple of S.Half; but if neg and pos are odd
+            # multiples of 1/2 then the difference will be an integer
+            # and that has already been handled with the check for n
+            # and k not integer but n - k is a negative integer
+            return
+
+        # k <= n < 0
+        if n.is_negative and k.is_negative:
+            if n_k.is_nonnegative:
+                if kint and nint:
+                    if not binom:
+                        return S.Zero
+                    n, k = -k - 1, n - k
+                    res = cls._eval(n, k)
+                    return res if res is None else Pow(S.NegativeOne, k)*res
+                if kint is False or nint is False:
+                    return cls._eval(n, k, 'gamma')
+                return
+
+        # 0 <= n < k
+        if n.is_nonnegative and k.is_positive:
+            if n_k.is_negative:
+                if nint and kint:
+                    return S.Zero
+                if n_kint:
+                    return S.Zero
+                if kint:
+                    return cls._eval(n, k)
+                if nint is False or kint is False:
+                    return cls._eval(n, k, 'gamma')
+                return
+
+        # k < 0 <= n
+        if n.is_extended_nonnegative and k.is_negative:
+            if kint:
+                return S.Zero
+            if n_kint:
+                return cls._eval(n, n_k)
+            if kint is False and n_kint is False:
+                return cls._eval(n, k, 'gamma')
+            return
+
+        # n < k < 0
+        if k.is_negative and n.is_negative:
+            if n_k.is_negative:
+                if n_kint:
+                    return S.Zero
+                # already handled kint and k < 0 above
+                # so only this is left:
+                if kint is False and n_kint is False:
+                    return cls._eval(n, k, 'gamma')
+                return
+
+        # other numbers
+        if k.is_number and n.is_number:
+            if kint:
+                return cls._eval(n, k)
+            elif kint is False:
+                return cls._eval(n, k ,'gamma')
+
+    def _eval_evalf(self, prec):
```

Comment:
Why is it that you are having to write this? Is it because of the 3rd argument? If so, perhaps it's better to keep the args as n and k and deal with the 3rd argument as a hint instead of an argument? The hint can be stored as `self.binom` so it is accessible, however.

Reply:
can you revert the last two commits and make those changes @smichr , I'm having trouble with making them. ☹️

Reply:
Maybe can't be done? Looks like Function doesn't like to have other keywords. Or Basic. Hmmm. Maybe we could handle it with assumptions on the variables. If `k.is_positive is not False` then assume it is positive. I'll try.

Reply:
If you get a chance to look at this, take a look at LambertW. `LambertW(x).args == (x,)` but a branch argument is supported:

```python
>>> w=LambertW
>>> w(3,1)
W_{1}(3)
>>> w(3,0)
W(3)
>>> _.args
(3,)
>>> w(3,-1)
W_{−1}(3)
>>> _.args
(3, -1)
```

Reply:
OK, see if you can "take it here" from my `multi` branch. It passed doctests, which is at least an encouraging start.

Reply:
It is passing all tests. But I would have to look over this carefully to figure out if this is actually right. e.g.
```python
>>> binomial(n,k,1).equals(binomial(n-1,k,1)+binomial(n-1,k-1,1))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "...\sympy\core\expr.py", line 753, in equals
    diff = factor_terms(simplify(self - other), radical=True)
  File "...\sympy\simplify\simplify.py", line 666, in simplify

    expr = combsimp(expr)
  File "...\sympy\simplify\combsimp.py", line 54, in combsimp
    expr = _gammasimp(expr, as_comb=True)
  File "...\sympy\simplify\gammasimp.py", line 111, in _gammas
imp
    expr = expr.replace(binomial, rule)
  File "...\sympy\core\basic.py", line 1508, in replace
    rv = bottom_up(self, rec_replace, atoms=True)
  File "...\sympy\simplify\simplify.py", line 1149, in bottom_
up
    args = tuple([bottom_up(a, F, atoms, nonbasic) for a in args])
  File "...\sympy\simplify\simplify.py", line 1149, in <listco
mp>
    args = tuple([bottom_up(a, F, atoms, nonbasic) for a in args])
  File "...\sympy\simplify\simplify.py", line 1149, in bottom_
up
    args = tuple([bottom_up(a, F, atoms, nonbasic) for a in args])
  File "...\sympy\simplify\simplify.py", line 1149, in <listco
mp>
    args = tuple([bottom_up(a, F, atoms, nonbasic) for a in args])
  File "...\sympy\simplify\simplify.py", line 1152, in bottom_
up
    rv = F(rv)
  File "...\sympy\core\basic.py", line 1489, in rec_replace
    new = _value(expr, result)
  File "...\sympy\core\basic.py", line 1437, in <lambda>
    _value = lambda expr, result: value(*expr.args)
TypeError: rule() takes 2 positional arguments but 3 were given

>>> binomial(-1,-2,1)
-1
>>> binomial(-1,-2)
-1  <----------- should be zero?
```

Reply:
@smichr so that turns out to be due to this condition we set, 
![Screen Shot 2020-05-29 at 4 02 18 PM](https://user-images.githubusercontent.com/28682735/83250528-e3c21c80-a1c5-11ea-9e2f-aba31a3f6194.png)
if `S.One in ...`

Reply:
changing that brings a whole slew of errors, I'll try to correct that

Reply:
also these work 
```python3
>>> binomial(-5,-5)
1
```

Reply:
I'm not sure that's "working". For the KFC-binomial, a negative int arg for either `n` or `k` gives a result of 0.

Reply:
Yeah right! 😄 , Although I'm working on something locally, it has to do with the gamma limit fixed recently ... If it fixes the concrete.rst i'll push a change!